### PR TITLE
Parsers rewrite and stuff

### PR DESCRIPTION
**Pull request by FireRedz**
## General
* mostly just making the code looks cool and some fixes.

## Parser
Rewrited: .osr Parser and Skin Parser
Refactored: .osu Parser

## Animations
Added: Animation/easings.py and Time-based (???) animation thing (for Fadeout and Resize) instead of cursed steps stuff
Changed: More accurate HitFadeOut???

### CODE REVIEW

File: osr2mp4/Parser/jsonparser.py [hunk 1]
```diff
@@ -1,12 +1,7 @@
...
-
 def read(filename):
-	fileopen = open(filename, "r", encoding="utf-8")
-	text = fileopen.read()
-	# text = text.replace("\\", "/")
-	data = json.loads(text)
-
-	fileopen.close()
+	with open(filename, 'r', encoding='utf-8') as file:
+		data = json.loads(file.read())
```
Comment (christopher-dG) [inline osr2mp4/Parser/jsonparser.py:5]:
```suggestion
		data = json.load(f)
```

File: osr2mp4/ImageProcess/Animation/alpha.py [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
+def img_fade(img: Image, start: float, end: float, step: float):
 	outputs = []
 	for x in np.arange(start, end, step):
 		im = imageproc.newalpha(img, x)
 		outputs.append(im)
 	return outputs
 
 
-def list_fade(img, start, end, step):
+def list_fade(img: [Image], start: float, end: float, step: float):
```
Comment (christopher-dG) [inline osr2mp4/ImageProcess/Animation/alpha.py:17]: I don't think Python type checkers know how to interpret `[Image]`, you're supposed to do `List[Image]` instead (`List` being from `typing`).
(this applied to a bunch of places in the code)

File: osr2mp4/ImageProcess/Animation/alpha.py [hunk 3]
```diff
@@ -39,4 +44,38 @@ def fadein(img, start, end, step):
...
+	return res
+	
+def list_fade_(imgs: [Image], start: float, end: float, duration: float, easing: easings):
+	res = []
+	delta = duration/len(imgs)
+	cur_time = 0
+	
+	for img in imgs:
+		fade_val = easing(cur_time, start, end, duration)
+		res += [imageproc.newalpha(img, fade_val)]
```
Comment (christopher-dG) [inline osr2mp4/ImageProcess/Animation/alpha.py:66]:
```suggestion
		res.append(imageproc.newalpha(img, fade_val))
```

File: osr2mp4/ImageProcess/Animation/easings.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+	Easing shit - FireRedz
+	copied from my storyboard renderer which is copied from some random gist
+	
+	t is the current time (or position) of the tween.
+	b is the beginning value of the property.
+	c is the change between the beginning and destination value of the property. | basically destination-start
+	d is the total time of the tween.
+'''
+
+def easeLinear(t, b, c, d):
```
Comment (christopher-dG) [inline osr2mp4/ImageProcess/Animation/easings.py:14]: snake_case ?

File: osr2mp4/ImageProcess/Animation/size.py [hunk 5]
```diff
@@ -44,3 +46,37 @@ def grow(img, start, end, step):
...
+	return res
+	
+def list_resize_(imgs: [Image], start: float, end: float, duration: float, easing: easings):
+	res = []
+	delta = duration/len(imgs)
+	cur_time = 0
+	
+	for img in imgs:
+		size_val = easing(cur_time, start, end, duration)
+		res += [imageproc.change_size(img, size_val, size_val)]
```
Comment (christopher-dG) [inline osr2mp4/ImageProcess/Animation/size.py:68]:
```suggestion
		res.append(imageproc.change_size(img, size_val, size_val))
```

File: osr2mp4/ImageProcess/Animation/size.py [hunk 6]
```diff
@@ -44,3 +46,37 @@ def grow(img, start, end, step):
...
 
+
+
+""" EPIC VERSION """
+def img_resize_(img: Image, start: float, end: float, delta: float, duration: float, easing: easings):
+	res = []
+	
+	for time in np.arange(0, duration, delta):
+		size_val = easing(time, start, end, duration)
+		res += [imageproc.change_size(img, size_val, size_val)]
```
Comment (christopher-dG) [inline osr2mp4/ImageProcess/Animation/size.py:57]:
```suggestion
		res.append(imageproc.change_size(img, size_val, size_val))
```

File: osr2mp4/Parser/osuparser.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 
 
 class Beatmap:
-	def __init__(self, info, scale=1, colors=None, mods=None, lazy=True):
+	def __init__(self, info: dict, scale: float = 1, colors: dict = None, mods: 'mods' = None, lazy: bool = True):
```
Comment (christopher-dG) [inline osr2mp4/Parser/osuparser.py:9]: `1` is not a float :-)

File: osr2mp4/Parser/osuparser.py [hunk 8]
```diff
@@ -62,107 +63,111 @@ def __init__(self, info, scale=1, colors=None, mods=None, lazy=True):
...
+		general = self.info["General"].split("\n")
 		for item in general:
 			item = item.strip()
 			if item != "":
-				my_list = item.split(":")
-				my_list[1] = my_list[1].strip()
-				self.general[my_list[0]] = float(my_list[1]) if my_list[1].replace('.', '', 1).isdigit() else my_list[1]
+				name, value = item.split(":")
+				value = value.strip()
+				self.general[name] = float(value) if (value.replace('.', '', 1)).isdigit() else value
```
Comment (christopher-dG) [inline osr2mp4/Parser/osuparser.py:72]:
```suggestion
				self.general[name] = float(value) if value.replace('.', '', 1).isdigit() else value
```

File: osr2mp4/Parser/osuparser.py [hunk 9]
```diff
@@ -62,107 +63,111 @@ def __init__(self, info, scale=1, colors=None, mods=None, lazy=True):
...
-		for item in general:
+		meta = self.info["Metadata"].split("\n")
+		for item in meta:
 			item = item.strip()
 			if item != "":
-				my_list = item.split(":")
-				my_list[1] = my_list[1].strip()
-				self.meta[my_list[0]] = float(my_list[1]) if my_list[1].replace('.', '', 1).isdigit() else my_list[1]
+				name, *value = item.split(":")
+				value = ' '.join(value).strip()
```
Comment (christopher-dG) [inline osr2mp4/Parser/osuparser.py:80]:
```suggestion
				name, value = item.strip().split(":", maxsplit=1)
```

File: osr2mp4/osrparse/replay.py [hunk 10]
```diff
@@ -14,74 +14,83 @@ def __init__(self, time_since_previous_action, x, y, keys_pressed):
...
+	score: int = 0
+	max_combo: int = 0
+	is_perfect_combo: int = 0
+	mod_combination: [Mod] = []
+	life_bar_graph: str = ''
+	timestamp: datetime.datetime = None
+	play_data: [ReplayEvent] = []
+
+	# internal
+	view: [memoryview, bytes] = None # memoryview or bytes
```
Comment (christopher-dG) [inline osr2mp4/osrparse/replay.py:46]: `Union[memoryview, bytes]`?

File: osr2mp4/osrparse/replay.py [hunk 11]
```diff
@@ -92,92 +101,92 @@ def bits(n):
 				b = n & (~n+1)
 				yield b
 				n ^= b
-
+
```
Comment (christopher-dG) [inline osr2mp4/osrparse/replay.py:104]:
```suggestion

```

File: osr2mp4/osrparse/replay.py [hunk 12]
```diff
@@ -92,92 +101,92 @@ def bits(n):
...
 
-	def parse_beatmap_hash(self, replay_data):
-		self.beatmap_hash = self.parse_string(replay_data)
+		return val
 
-	def parse_replay_hash(self, replay_data):
-		self.replay_hash = self.parse_string(replay_data)
+	def read_string(self):
+		if self.read_byte() == 0x00:
+			return ''
```
Comment (christopher-dG) [inline osr2mp4/osrparse/replay.py:170]:
```suggestion
			return ""
```

The existing code seems to use double quotes, I'd recommend trying to stay consistent with that throughout

**Review by christopher-dG — COMMENTED** [review]
Hope ya don't mind some random suggestions :-)(

File: osr2mp4/ImageProcess/Animation/alpha.py (hunk 2 above)
Comment (FireRedz) [reply osr2mp4/ImageProcess/Animation/alpha.py:17]: aight

File: osr2mp4/ImageProcess/Animation/easings.py (hunk 4 above)
Comment (FireRedz) [reply osr2mp4/ImageProcess/Animation/easings.py:14]: nah i prefer it like this, might change this in the future

File: osr2mp4/osrparse/replay.py (hunk 10 above)
Comment (FireRedz) [reply osr2mp4/osrparse/replay.py:46]: havent tried the typing stuff yet, will change this

File: osr2mp4/osrparse/replay.py (hunk 12 above)
Comment (FireRedz) [reply osr2mp4/osrparse/replay.py:169]: noted

File: osr2mp4/ImageProcess/Animation/alpha.py (hunk 3 above)
Comment (FireRedz) [reply osr2mp4/ImageProcess/Animation/alpha.py]: yea so about this, idk why i started doing the `list += [val]` thing. somehow it just works in my brain

File: osr2mp4/Parser/osuparser.py (hunk 8 above)
Comment (FireRedz) [reply osr2mp4/Parser/osuparser.py:72]: nah

**Comment by FireRedz** [comment]
ok this is going to take awhile, turns out school is a thing now.

**Comment by FireRedz** [comment]
been using this for over 2 weeks and nothing gone wrong yet so i think it should be stable enough.
will do another pr for another stuff

**Comment by rosenishere** [comment]
amogus

**Comment by uyitroa** [comment]
sus

**Pull request merged**